src/sinks/util/http.rs

```diff
@@ -297,27 +306,39 @@ impl<B> HttpBatchService<B> {
     }
 }
 
-impl<B> Service<B> for HttpBatchService<B> {
-    type Response = Response;
-    type Error = Error;
+impl<F, B> Service<B> for HttpBatchService<F, B>
+where
+    F: Future<Output = crate::Result<hyper::Request<Vec<u8>>>> + Send + 'static,
+    B: Send + 'static,
+{
+    type Response = http02::Response<Bytes>;
+    type Error = crate::Error;
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Ok(()).into()
+        Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, body: B) -> Self::Future {
-        let request = (self.request_builder)(body).map(Body::from);
+        let request_builder = Arc::clone(&self.request_builder);
+        let mut http_client = self.inner.clone();
 
-        let response = self.inner.call(request);
-        let fut = async move {
-            let res = response.await?;
-            let (parts, body) = res.into_parts();
+        Box::pin(async move {
+            let request = request_builder(body).await?.map(Body::from);
+            let response = http_client.call(request).await?;
+            let (parts, body) = response.into_parts();
             let mut body = body::aggregate(body).await?;
             Ok(hyper::Response::from_parts(parts, body.to_bytes()))
-        };
+        })
+    }
+}
 
-        Box::pin(fut)
+impl<F, B> Clone for HttpBatchService<F, B> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            request_builder: self.request_builder.clone(),
+        }
     }
 }
```

Comment:
This looks like what `#[derive(Clone)]` would produce. Was there a problem with the standard derivation?

Reply:
With derive I get error:
```
error[E0277]: the trait bound `(dyn core::future::future::Future<Output = std::result::Result<http::request::Request<std::vec::Vec<u8>>, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static): std::clone::Clone` is not satisfied
   --> src/sinks/util/http.rs:127:38
    |
127 |         let inner = request_settings.batch_sink(logic, svc, batch, batch_settings, cx.acker());
    |                                      ^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `(dyn core::future::future::Future<Output = std::result::Result<http::request::Request<std::vec::Vec<u8>>, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)`
    |
    = note: required because of the requirements on the impl of `std::clone::Clone` for `std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<http::request::Request<std::vec::Vec<u8>>, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)>`
    = note: required because of the requirements on the impl of `std::clone::Clone` for `std::pin::Pin<std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<http::request::Request<std::vec::Vec<u8>>, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)>>`
    = note: required because of the requirements on the impl of `std::clone::Clone` for `sinks::util::http::HttpBatchService<std::pin::Pin<std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<http::request::Request<std::vec::Vec<u8>>, std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + std::marker::Send + 'static)>>, <B as sinks::util::batch::Batch>::Output>`
```

`HttpClient` also implement `Clone` instead derive =\\
https://github.com/timberio/vector/blob/c293c492e97a7249822be4907f6bab84414dae7d/src/sinks/util/http.rs#L259-L267

Reply:
I do not know why I get such errors, `HttpClient` and `Arc` implements `Clone`. Will happy change to derive if somebody know solution.

Reply:
I see. I tried adding `F: Clone` trait bounds to the struct, but that didn't help neither. Oh well, not a big deal.

Reply:
It's wired. I'd like dig into this problem for understanding Rust more, but not sure that this cost it in current PR and work time.